lib/assert.js

```diff
@@ -369,9 +382,159 @@ function isSpecial(obj) {
+    if (expectedViewLength > actualByteLength) {
+      return false;
+    }
+
+    actualView = new Uint8Array(
+      DataViewPrototypeGetBuffer(actual),
+      DataViewPrototypeGetByteOffset(actual),
+      actualByteLength,
+    );
+    expectedView = new Uint8Array(
+      DataViewPrototypeGetBuffer(expected),
+      DataViewPrototypeGetByteOffset(expected),
+      expectedViewLength,
+    );
+  } else {
+    if (ObjectPrototypeToString(actual) !== ObjectPrototypeToString(expected)) {
+      return false;
+    }
+    actualView = actual;
+    expectedView = expected;
+    expectedViewLength = TypedArrayPrototypeGetLength(expected);
+
+    if (expectedViewLength > TypedArrayPrototypeGetLength(actual)) {
+      return false;
+    }
+  }
+
+  for (let i = 0; i < expectedViewLength; i++) {
+    if (actualView[i] !== expectedView[i]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function compareSets(actual, expected, comparedObjects) {
+  if (expected.size > actual.size) {
+    return false; // `expected` can't be a subset if it has more elements
+  }
+
+  if (isDeepEqual === undefined) lazyLoadComparison();
+
+  const actualArray = ArrayFrom(FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual));
+  const expectedIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], expected);
+  const usedIndices = new SafeSet();
+
+  expectedIteration: for (const expectedItem of expectedIterator) {
+    for (let actualIdx = 0; actualIdx < actualArray.length; actualIdx++) {
+      if (!usedIndices.has(actualIdx) && isDeepStrictEqual(actualArray[actualIdx], expectedItem)) {
+        usedIndices.add(actualIdx);
+        continue expectedIteration;
+      }
+    }
+    return false;
+  }
+
+  return true;
+}
+
+function compareArrays(actual, expected, comparedObjects) {
+  if (expected.length > actual.length) {
+    return false;
+  }
+
+  if (isDeepEqual === undefined) lazyLoadComparison();
+
+  // Create a map to count occurrences of each element in the expected array
+  const expectedCounts = new SafeMap();
+  for (const expectedItem of expected) {
+    let found = false;
+    for (const { 0: key, 1: count } of expectedCounts) {
+      if (isDeepStrictEqual(key, expectedItem)) {
+        MapPrototypeSet(expectedCounts, key, count + 1);
+        found = true;
+        break;
+      }
+    }
+    if (!found) {
+      MapPrototypeSet(expectedCounts, expectedItem, 1);
+    }
+  }
+
+  const safeActual = new SafeArrayIterator(actual);
+
+  // Create a map to count occurrences of relevant elements in the actual array
+  for (const actualItem of safeActual) {
+    for (const { 0: key, 1: count } of expectedCounts) {
+      if (isDeepStrictEqual(key, actualItem)) {
+        if (count === 1) {
+          MapPrototypeDelete(expectedCounts, key);
+        } else {
+          MapPrototypeSet(expectedCounts, key, count - 1);
+        }
+        break;
+      }
+    }
+  }
+
+  return !expectedCounts.size;
```

Comment:
We should probably clear the map to ensure we're not leaking memory here
```suggestion
  const { size } = expectedCounts;
  expectedCounts.clear();
  return size === 0;
```